src/lib/rule2F.ts

```diff
@@ -138,6 +138,93 @@ function allowsNameFromContent(elem: HTMLElement): boolean {
 
 export const TEST_ONLY = {allowsNameFromContent};
 
+// See https://lists.w3.org/Archives/Public/public-aria/2017Jun/0057.html
+// for discussion of roles & tags that forbid name from content.
+const NO_NAME_FROM_CONTENT_ROLES = [
+  'application',
+  'alert',
+  'log',
+  'marquee',
+  'timer',
+  'alertdialog',
+  'dialog',
+  'banner',
+  'complementary',
+  'form',
+  'main',
+  'navigation',
+  'region',
+  'search',
+  'article',
+  'document',
+  'feed',
+  'figure',
+  'img',
+  'math',
+  'toolbar',
+  'menu',
+  'menubar',
+  'grid',
+  'listbox',
+  'radiogroup',
+  'textbox',
+  'searchbox',
+  'spinbutton',
+  'scrollbar',
+  'slider',
+  'tablist',
+  'tabpanel',
+  'tree',
+  'treegrid',
+  'separator',
+  'rowgroup',
+  'group',
+];
+const NO_NAME_FROM_CONTENT_TAGS = [
+  'article',
+  'aside',
+  'body',
+  'select',
+  'datalist',
+  'optgroup',
+  'dialog',
+  'figure',
+  'footer',
+  'form',
+  'header',
+  'hr',
+  'img',
+  'textarea',
+  'input',
+  'main',
+  'math',
+  'menu',
+  'nav',
+  'section',
+  'thead',
+  'tbody',
+  'tfoot',
+  'fieldset',
+];
+
+/**
+ * Checks if 'elem' is forbidden from allowing 'name from content'
+ * @param elem - element whose text alternative is being computed
+ */
+function forbidsNameFromContent(elem: HTMLElement): boolean {
+  const explicitRole = elem.getAttribute('role')?.trim().toLowerCase() ?? '';
```

Comment:
Should this also include implicit roles?

Reply:
The implicit roles are handled in the next `if` statement:

```
const elemNodeName = elem.nodeName.toLowerCase();
  if (NEVER_NAME_FROM_CONTENT_TAGS.includes(elemNodeName)) {
    return true;
  }
```

I'll add a comment here to make this clear and more readable.

Reply:
But the tag name doesn't entirely represent the semantics of  role. For example `<header>` only has a role of `banner` if it's `not a descendant of an article, aside, main, nav or section element, or an element with role=article, complementary, main, navigation or region`... https://www.w3.org/TR/html-aria/#docconformance

For now please just raise an issue, I can add role-finding once my code for that is open source